remote-vfs.c

```diff
@@ -0,0 +1,1889 @@
+	strbuf_addstr(&idx_name_dst, ".idx");
+
+	// TODO if either pack_name_dst or idx_name_dst already
+	// TODO exists in the ODB, create alternate names so that
+	// TODO we don't step on them.
+
+	if (rename_tempfile(pp_tempfile, pack_name_tmp.buf) == -1)
+		die("could not rename packfile to '%s'", pack_name_tmp.buf);
+
+	argv_array_push(&ip.args, "index-pack");
+	argv_array_pushl(&ip.args, "-o", idx_name_tmp.buf, NULL);
+	argv_array_push(&ip.args, pack_name_tmp.buf);
+	ip.git_cmd = 1;
+	ip.no_stdin = 1;
+	ip.no_stdout = 1;
+
+	// TODO consider capturing stdout from index-pack because
+	// TODO it will contain the SHA of the packfile and we can
+	// TODO (should?) add it to the .pack and .idx pathnames
+	// TODO when we install them.
+	// TODO
+	// TODO Or should be SHA-it ourselves (or read the last 20 bytes)?
+
+	if (run_command(&ip)) {
+		unlink(pack_name_tmp.buf);
+		unlink(idx_name_tmp.buf);
+		die("index-pack failed on '%s'", pack_name_tmp.buf);
+	}
+
+	if (finalize_object_file(pack_name_tmp.buf, pack_name_dst.buf) ||
+	    finalize_object_file(idx_name_tmp.buf, idx_name_dst.buf)) {
+		unlink(pack_name_tmp.buf);
+		unlink(pack_name_dst.buf);
+		unlink(idx_name_tmp.buf);
+		unlink(idx_name_dst.buf);
+		die("could not install packfile '%s'", pack_name_dst.buf);
+	}
+
+	strbuf_setlen(packfile_name, 0);
+	strbuf_addbuf(packfile_name, &pack_name_dst);
+
+	child_process_clear(&ip);
+	strbuf_release(&pack_name_tmp);
+	strbuf_release(&pack_name_dst);
+	strbuf_release(&idx_name_tmp);
+	strbuf_release(&idx_name_dst);
+}
+
+/*
+ * Convert the tempfile into a permanent loose object in the ODB.
+ *
+ * Return the full pathname of the resulting file.
+ */
+static void install_loose(struct tempfile **pp_tempfile,
+			  struct strbuf *loose_path)
+{
+	struct strbuf loose_path_tmp = STRBUF_INIT;
+	size_t len_base;
+
+	/*
+	 * start with "<odb>/xx/y38.temp" and owned by tempfile class.
+	 * compute new name as "<odb>/xx/y38".
+	 * close tempfile to steal ownership.
+	 * officially install loose object.
+	 */
+
+	strbuf_addstr(&loose_path_tmp, get_tempfile_path(*pp_tempfile));
+	close_tempfile_gently(*pp_tempfile);
+	pp_tempfile = NULL;
+
+	if (!strip_suffix(loose_path_tmp.buf, ".temp", &len_base))
+		BUG("loose object tempfile does not end in '.temp': '%s'",
+		    loose_path_tmp.buf);
+
+	strbuf_setlen(loose_path, 0);
+	strbuf_add(loose_path, loose_path_tmp.buf, len_base);
+
+	if (finalize_object_file(loose_path_tmp.buf, loose_path->buf)) {
+		unlink(loose_path_tmp.buf);
+		die("could not install loose object '%s'", loose_path->buf);
+	}
+
+	strbuf_release(&loose_path_tmp);
+}
+
+//////////////////////////////////////////////////////////////////
+
+/*
+ * Our wrapper to initialize the HTTP layer.
+ *
+ * We always use the real origin server, not the cache-server.
+ */
+static void rv_http_init(void)
+{
+	if (rv_data.http_is_initialized)
+		return;
+
+	http_init(rv_data.remote, rv_data.buf_main_url.buf, 0);
+	rv_data.http_is_initialized = 1;
+}
```

Comment:
Hopefully we never call this in concurrent threads...

Reply:
Is the http.c code even thread safe?  there are static vars all over the place.  especially all the creds.